rfcs/0171-fetch-from-github.md

```diff
@@ -0,0 +1,115 @@
+---
+feature: Default name of fetchFromGithub FOD to include revision
+start-date: 2024-03-15
+author: Jonathan Ringer
+co-authors: 
+shepherd-team:  
+shepherd-leader: 
+related-issues:
+---
+
+# Summary
+[summary]: #summary
+
+Updating the hash on Fixed-Output Derivations (FODs) is a very error prone process. It's not intuitive to invalidate the existing hash, attempt to realize the FOD, then replace the hash value with the value that Nix just calculated. This RFC advocates for influencing the default derivation name of the fetchFromGithub helper with the "repo" and "rev" values to ensure that changed URLs force the FOD to be re-built.
+
+# Motivation
+[motivation]: #motivation
+
+This will hopefully provide immediate feedback that an FOD contains a stale hash. One must either build the FOD without the previous FOD in their Nix store, or run the FOD build with `--check` to verify that the FOD is not stale. Although fetchFromGithub is one of many fetchers; it is very common, and generally has a user specify granular source information which makes differentiating between sources easy.
+
+As a secondary effect, this will also give a more meaningful name to the FODs than "source". E.g. `/nix/store/...-source -> /nix/store/...-protobuf-v24.1-src`
+
+# Detailed design
+[design]: #detailed-design
+
+Change the default name of fetchFromGithub fetcher from `"source"` to `lib.sanitizeDerivationName "${repo}-${rev}-src"`.
```

Comment:
> How about using pname instead of repo as a sort of middle ground between stability and informative names. Of course then changing the repo url will not automaticaly verify if the hash actually matches, so a tradeoff.

`pname` is a stdenv.mkDerivation convention, this would require people to pass it separately. The goal here also is to avoid hash mismatch, so I don't think is this is enough to solve the problem this RFC is trying to remedy.

> Also how about starting with src- instead of at the end? When listing the nix store, all fetched inputs can then be sorted together.

I wouldn't be opposed to it, but I'm used to `<pname>-<version>-<variant>` naming, source being a variant in this case.

Reply:
> The second suggestion is actually a part of a wish I have to prepend src- to all build inputs (tars, patch files, builders, etc), pkg- to all packages, and nixos- to the generated nixos configs, so that it is easier to tell at a glance why a particular store path is being installed, and also would be useful for tools like nvd when listing diffs. Yeah this would a much broader proposal, but one can dream someday it will happen. For now, we can start with just the FOD fetchers.

It sounds like another Nix RFC.

Reply:
> I'm used to `<pname>-<version>-<variant>` naming, source being a variant in this case.

It's a bit strange to consider package source a "variant" of the package. Nevertheless, such name-suffixing way could already be found in Nixpkgs -- `goModules` used by `buildGoModule` does use the suffixing `name` pattern, and sometimes people manually set the in-package FODs to name the suffixed way.

On the other hand, language-specific packages (e.g. Python packages, Ocaml packages, etc.) often prefix the name with the name and version of the compiler/interpreter.

Reply:
I'm not specifically tied to the suffix approach.. I'll add a comment to the initial comment